src/Pact/Types/Runtime.hs

```diff
@@ -426,10 +441,40 @@ data Purity =
 instance Default Purity where def = PImpure
 
 -- | Marker class for 'PNoDb' environments.
-class PureNoDb e where
+class PureNoDb e
 -- | Marker class for 'PSysRead' environments.
 -- SysRead supports pure operations as well.
-class PureNoDb e => PureSysRead e where
+class PureNoDb e => PureSysRead e
+
+-- | DB Read value for per-row gas costing.
+-- Data is included if variable-size.
+data ReadValue
+  = ReadData (Columns Persistable)
+  | ReadKey RowKey
+  | ReadTxId
+
+
+data GasArgs
+  = GPostRead ReadValue
+  | GSelect (Maybe [(Info,ColumnId)]) (Term Ref) (Term Name)
+  | GUnreduced [Term Ref]
+  | GReduced [Term Name]
+  | GUse ModuleName (Maybe Hash)
+  | GModule Module
+  | GModuleMember Module
+  | GUser
+
+
+newtype GasModel = GasModel { runGasModel :: Text -> GasArgs -> Gas }
```

Comment:
`Text` is the operation name, see the PR comment about matching on text.

A non-const gas model would match on the op name and the args data. For something simple like `+` this would most likely return a fixed cost. For `write` we might want to evaluate the update data to compute a size-based metric.

```haskell
case (name,args) of
  ("+",_) -> return BINOP_PRICE -- some constant price for math binops
  ("write",[TObject {..}]) -> return WRITE_PRICE + computeWriteSize _tPairs 
  ...
```